### DISCUSSION TITLE
Integration test across the bridge

### DISCUSSION THREAD

**Discussion by sundeepgupta** [ionic-team/capacitor#3933] [Q&A]
I was wondering if anyone has done this sort of thing already and could explain how it might work. 

What I'd like to know is how I could write a test in the web code (JS) that would use native code (iOS in this case is fine). And vice-versa, so that a test in the iOS code that would use JS code. 

I'm an iOS developer and I'm wondering if I'd need to setup a duplicate (or perhaps modified for test) Capacitor bridge and web app for the test target for this to work. 

Furthermore, how could each platform's test code could control the other platform's dependencies and mocks/fakes/doubles/fixtures/etc.

**Comment by pwespi** [comment]
@sundeepgupta Have you found a solution?

I'm also wondering how to best test a Capacitor plugin. My current approach is to have an example app with a button that when clicked uses all the APIs of my plugin, checks the results and uses console.error when the result is not what expected. This way I can make sure that the plugin behaves the same on Android, iOS, and web and I run this manually before every release of the plugin.

I would be curious to hear if there is a better way to do it.

**Reply by sundeepgupta** [reply]
From my investigation into this, there is quite a bit of effort needed to make a good test workflow across the bridge. For one thing everything is async across the bridge so that makes the test code less than ideal. The other issue is that when going from native (iOS anyways) to JS, the JS cannot return a value. JS has to basically call into a plugin, which is out of context of the original native call. Kind of like [Apple's Delegate pattern](https://www.swiftbysundell.com/articles/delegation-in-swift/). 

Console logging isn't really great because the test runner can't tell you if the test suite passed or failed. 

Ultimately, we haven't implemented anything yet, but my thoughts on this are to not do it unless the value of the test is very high.
